core/src/main/java/org/testcontainers/lifecycle/Startables.java

```diff
@@ -72,6 +73,18 @@ public Thread newThread(Runnable r) {
             })
             .toArray(CompletableFuture[]::new);
 
-        return CompletableFuture.allOf(futures);
+        return allOfShortcircuiting(futures);
+    }
+
+    private static CompletableFuture<Void> allOfShortcircuiting(CompletableFuture[] futures) {
+        CompletableFuture<Void> result = new CompletableFuture<>();
+
+        for (CompletableFuture<?> f : futures) {
+            f.handle((__, ex) -> ex == null || result.completeExceptionally(ex));
+        }
+
+        CompletableFuture.allOf(futures).thenRun(() -> result.complete(null));
+
+        return result;
```

Comment:
there is a bit of ceremony involved in the `CompletableFuture` protocol, including the cancellation.
Please make sure that if the returned future is canceled, the inner futures are canceled too.

Reply:
also, you could use `return anyOf(allOf(futures), result)` where `result` resolves on failure

Reply:
> Please make sure that if the returned future is canceled, the inner futures are canceled too.

I'm not touching any of the original futures, I just attach a post-complete action for cases where futures are already failed

Reply:
> also, you could use return anyOf(allOf(futures), result) where result resolves on failure

In this case, it doesn't really make sense because the future works just as a signaller and just returns `Void`

Reply:
> I'm not touching any of the original futures

this is not the problem. The problem I see is that when the resulting future receives `CompletableFuture#cancel(boolean)`, ideally, we should also cancel the inner operations.

Reply:
> In this case, it doesn't really make sense because the future works just as a signaller and just returns Void

that's exactly the signature we have.

Reply:
> this is not the problem. The problem I see is that when the resulting future receives CompletableFuture#cancel(boolean), ideally, we should also cancel the inner operations.

`CompletableFuture#cancel(boolean)` cancels only the future itself, but doesn't propagate the interrupt as good-old `FutureTask`

Reply:
> > In this case, it doesn't really make sense because the future works just as a signaller and just returns Void
> 
> that's exactly the signature we have.

So you'd like to end up with something like `CompletableFuture<Void> f = anyOf(allOf(futures), null)`?

Reply:
No. As I said, I was willing to see `anyOf(allOf(futures), result)` where result resolves only on failure.

Reply:
Ok, I think I got what you meant and it took me a while...  because the code you commented is already outdated. 

I managed to avoid the extra cf by introducing a simple race:

```
private static CompletableFuture<Void> firstFailing(CompletableFuture[] futures) {
        CompletableFuture<Void> result = CompletableFuture.allOf(futures);

        for (CompletableFuture<?> f : futures) {
            f.whenComplete((__, ex) -> {
                if (ex != null) {
                    result.completeExceptionally(ex);
                }
            });
        }

        return result;
    }
```